### PR TITLE
UnityCatalog-ai docs

### PR DESCRIPTION
**Pull request by BenWilson2**
**PR Checklist**

- [X] A description of the changes is added to the description of this PR.
- [ ] If there is a related issue, make sure it is linked to this PR.
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added or modified a feature, documentation in `docs` is updated

**Description of changes**

This is the initial PR 1/x for the UnityCatalog-ai library. 
Currently, while we work on the OSS client, there are placeholders for the OSS Client docs (these will be addressed in a follow-up doc), along with additional guidance on caveats for its configuration and usage. 
Once the OSS Client is available, we will update all integration examples and documentation to show OSS client first, then Databricks Client in the top portion of the tutorials.

### CODE REVIEW

**Comment by BenWilson2** [comment]
cc @serena-ruan for first-pass review

File: docs/ucai/index.md [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+
+- **Dependency Management**
+
+    - Pitfall: Missing dependencies like pandas can cause runtime errors when executing functions that require them.
+    - Solution: Ensure all necessary dependencies are installed in the environment where the client operates. Use virtual environments to manage dependencies cleanly.
+
+- **Compatibility with Databricks Connect**
+
+    - Pitfall: Using incompatible versions of databricks-connect can lead to execution failures, especially with serverless compute.
+    - Solution: Install the specific version of databricks-connect that is supported (15.1.0 or higher) to ensure compatibility with serverless features.
```
Comment (serena-ruan) [inline docs/ucai/index.md:52]:
```suggestion
    - Solution: Install the specific version of databricks-connect that is supported (15.1.0) to ensure compatibility with serverless features.
```

File: docs/ucai/index.md [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+```
+
+If you decide to run operations using serverless runtime on Databricks, you do not need to define a `warehouse_id` when creating the client.
+
+#### Gotchas and Best Practices for the Client
+
+- **Warehouse ID Accuracy**
+
+    - Pitfall: Providing an incorrect warehouse_id can lead to authentication failures or inability to execute functions.
+    - Solution: Double-check the warehouse_id obtained from your Databricks workspace. Ensure that the warehouse is active and supports serverless compute if required.
```
Comment (serena-ruan) [inline docs/ucai/index.md:27]:
```suggestion
    - Solution: Double-check the warehouse_id obtained from your Databricks workspace. Ensure that the warehouse is active and of type `serverless`.
```

File: docs/ucai/index.md [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+
+CATALOG = "my_catalog"
+SCHEMA = "my_schema"
+```
+
+If you decide to run operations using serverless runtime on Databricks, you do not need to define a `warehouse_id` when creating the client.
+
+#### Gotchas and Best Practices for the Client
+
+- **Warehouse ID Accuracy**
```
Comment (serena-ruan) [inline docs/ucai/index.md:24]: Do you think table is better here for displaying?

File: docs/ucai/index.md [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+``` python
+from ucai.core.databricks import DatabricksFunctionClient
+
+# Initialize the client using serverless
+client = DatabricksFunctionClient()
+
+# Define your catalog and schema
+CATALOG = "my_catalog"
+SCHEMA = "my_schema"
+
```
Comment (serena-ruan) [inline docs/ucai/index.md:66]: I think we can remove this part, as the client is defined above

File: docs/ucai/index.md [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+- **Type Hint Requirements**
+    - All Python functions passed to `create_python_function` must have explicit type hints for both parameters and return values. These type hints are used to determine the corresponding SQL data types for the UC function.
+    - If type hints are missing, the function will not be accepted.
+   
+- **Supported Python to SQL Type Mapping**
+    - `int` → `LONG`
+    - `float` → `DOUBLE`
+    - `str` → `STRING`
+    - `bool` → `BOOLEAN`
+    - `Decimal` → `DECIMAL(38,18)` (with default precision and scale)
```
Comment (serena-ruan) [inline docs/ucai/index.md:92]: I thought default is Decimal(10, 0) https://docs.databricks.com/en/sql/language-manual/data-types/decimal-type.html#syntax

File: docs/ucai/index.md [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+executed prior to proceeding with using it in a GenAI solution to reduce the amount of troubleshooting you may have to do.
+
+
+
+### Using a SQL body statement
+
+An alternative to using the `create_python_function` API is to use the `create_function` API. This method requires you to submit the full SQL body for the function
+creation manually.
+
+For example:
```
Comment (serena-ruan) [inline docs/ucai/index.md:166]: This is duplicate from client.md, maybe we just link to it?

File: docs/ucai/index.md [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+
+- **Permissions and Access Control**
+
+    - Pitfall: Insufficient permissions can prevent function creation or execution.
+    - Solution: Ensure that the user account initializing the DatabricksFunctionClient has the necessary permissions to create and manage functions within the specified catalog and schema.
+
+- **Function Overwriting**
+
+    - Pitfall: Attempting to create a function that already exists without specifying the replace parameter can fail.
+    - Solution: Include OR REPLACE in your SQL statement if you intend to overwrite an existing function.
```
Comment (serena-ruan) [inline docs/ucai/index.md:210]:
```suggestion
    - Solution: Include `OR REPLACE` in your SQL statement if you intend to overwrite an existing function.
```

File: docs/ucai/index.md [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+| `int`                    | `LONG`                      | Represents large integer values.                                            |
+| `float`                  | `DOUBLE`                    | Represents double-precision floating-point values.                          |
+| `str`                    | `STRING`                    | Represents variable-length string data.                                     |
+| `bool`                   | `BOOLEAN`                   | Represents boolean values (`True`/`False`).                                 |
+| `datetime.date`          | `DATE`                      | Represents date values (year, month, day).                                  |
+| `datetime.datetime`      | `TIMESTAMP`                 | Represents timestamp values (date and time).                                |
+| `datetime.timedelta`     | `INTERVAL DAY TO SECOND`    | Represents intervals (difference between two timestamps).                   |
+| `decimal.Decimal`        | `DECIMAL(38, 18)`           | Represents fixed-point decimal values with default precision and scale.     |
```
Comment (serena-ruan) [inline docs/ucai/index.md:249]: Let's link line 87 here instead of duplicating it twice?

File: docs/ucai/integrations/index.md [hunk 9]
```diff
@@ -0,0 +1,17 @@
+# Unity Catalog AI Integrations
+
+## LangChain
+
```
Comment (serena-ruan) [inline docs/ucai/integrations/index.md:4]: TODO: fill links

File: docs/ucai/integrations/langchain.md [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+    ]
+)
+
+# Define the agent, specifying the tools from the toolkit above
+agent = create_tool_calling_agent(llm, tools, prompt)
+
+# Create the agent executor
+agent_executor = AgentExecutor(agent=agent, tools=tools, verbose=True)
+agent_executor.invoke({"input": "What is 36939 * 8922.4?"})
+```
```
Comment (serena-ruan) [inline docs/ucai/integrations/langchain.md:130]: Once we have notebooks for each integration package, let's link them at the end of this markdown file? e.g. https://github.com/unitycatalog/unitycatalog/blob/main/unitycatalog-ai/integrations/langchain/ucai-langchain_sample.ipynb

**Review by serena-ruan — COMMENTED** [review]
TODO: rename package names :D

File: docs/ucai/index.md (hunk 3 above)
Comment (BenWilson2) [reply docs/ucai/index.md:24]: yep!

File: docs/ucai/index.md (hunk 5 above)
Comment (BenWilson2) [reply docs/ucai/index.md:92]: ah I forgot about that. We can't use the default though since it's lossy for large values. I map to the max scale to avoid truncation / rounding issues.

File: docs/ucai/index.md (hunk 8 above)
Comment (BenWilson2) [reply docs/ucai/index.md:249]: I simply forgot to delete this placeholder :)

File: ai/core/src/unitycatalog/ai/core/utils/type_utils.py [hunk 11]
```diff
@@ -10,7 +10,7 @@
     datetime.date: "DATE",
     datetime.datetime: "TIMESTAMP",
     datetime.timedelta: "INTERVAL DAY TO SECOND",
-    decimal.Decimal: "DECIMAL(38, 18)",  # default precision and scale
+    decimal.Decimal: "DECIMAL(38, 18)",
```
Comment (serena-ruan) [inline ai/core/src/unitycatalog/ai/core/utils/type_utils.py:13]: Can we revert this change, just in case we forgot why we put 38 and 18 here :)

File: docs/ai/client.md [hunk 12]
```diff
@@ -0,0 +1,256 @@
+# Client Guide
+
+This guide provides detailed information about the Unity Catalog AI clients, including caveats, environment variables, public APIs, and examples to help you effectively utilize the Unity Catalog AI Core Library.
+
+---
+
+## Open Source Function Client
+
```
Comment (serena-ruan) [inline docs/ai/client.md:8]: Maybe add a message or placeholder here in case we forget to update?

File: docs/ai/integrations/openai.md [hunk 13]
```diff
@@ -0,0 +1,176 @@
+# 🦙 Using Unity Catalog AI with the OpenAI SDK
```
Comment (serena-ruan) [inline docs/ai/integrations/openai.md:1]:
```suggestion
# Using Unity Catalog AI with the OpenAI SDK
```

File: docs/ai/integrations/anthropic.md [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+### Unity Catalog Open Source
+
+Ensure that you have a functional UC server set up and that you are able to access the catalog and schema where defined functions are stored.
+
+### Databricks Unity Catalog
+
+To interact with Databricks Unity Catalog, ensure that you have both the `databricks-sdk` and the `databricks-connect` packages installed:
+
+```sh
+pip install databricks-sdk "databricks-connect>=15.1.0"
```
Comment (serena-ruan) [inline docs/ai/integrations/anthropic.md:36]: I think we should provide `unitycatalog-anthropic[databricks]` instead of asking them to install all extra dependencies needed for databricks usage. `unitycatalog-ai[databricks]` is already existing

File: docs/ai/quickstart.md [hunk 15]
```diff
@@ -0,0 +1,170 @@
+# Quickstart Guide
+
+Welcome to the **Unity Catalog AI Core Library**! This guide will help you get started with installing the library, setting up your environment, and creating and executing Unity Catalog (UC) functions. Whether you're using your own UC open-source server or connecting with a Databricks-managed UC, this quickstart will walk you through the essential steps.
```
Comment (serena-ruan) [inline docs/ai/quickstart.md:3]:
```suggestion
Welcome to the **Unity Catalog AI Core Library**! This guide will help you get started with installing the library, setting up your environment, creating and executing Unity Catalog (UC) functions. Whether you're using your own UC open-source server or connecting with a Databricks-managed UC, this quickstart will walk you through the essential steps.
```

File: docs/ai/quickstart.md [hunk 16]
```diff
@@ -0,0 +1,170 @@
...
+If you're looking to use an Open Source Unity Catalog service, ensure that you have met all of the prerequisites and have followed the
+[Unity Catalog Quickstart Guide](../quickstart.md). Once your server is up and running and you are able to create a catalog, schemas, and functions, you're all set to continue with this guide.
+
+### Databricks Unity Catalog
+
+If you are going to interface with ``Databricks Unity Catalog``, ensure that you have installed the ``Databricks SDK`` to enable access
+to your Databricks Unity Catalog service.
+
+```sh
+pip install databricks-sdk
```
Comment (serena-ruan) [inline docs/ai/quickstart.md:32]:
```suggestion
pip install unitycatalog-ai[databricks]
```
we could mention what extra packages are installed by this, but using `[databricks]` can save us some maintenance work in case we update the dependencies in the future.

**Review by serena-ruan — APPROVED** [review]
Overall LGTM! Awesome docs!

File: docs/ai/integrations/openai.md (hunk 13 above)
Comment (BenWilson2) [reply docs/ai/integrations/openai.md:1]: LOL whoops

File: docs/ai/integrations/anthropic.md (hunk 14 above)
Comment (BenWilson2) [reply docs/ai/integrations/anthropic.md:36]: great idea! I'll file another PR to update all of our integration packages to allow for these optional imports.

Comment (BenWilson2) [reply docs/ai/integrations/anthropic.md:36]: https://github.com/unitycatalog/unitycatalog/pull/737
